File: src/main/java/sk/hudak/prco/ssl/PrcoCustomKeyManager.java
package sk.hudak.prco.ssl;

import lombok.extern.slf4j.Slf4j;

import javax.net.ssl.KeyManager;
import javax.net.ssl.KeyManagerFactory;
import javax.net.ssl.X509KeyManager;
import java.net.Socket;
import java.security.Principal;
import java.security.PrivateKey;
import java.security.cert.X509Certificate;

@Slf4j
public class PrcoCustomKeyManager implements X509KeyManager {


    private X509KeyManager javaDefaultKeyManager;

    public PrcoCustomKeyManager() throws Exception {
        log.debug("inicializing");
        this.javaDefaultKeyManager = (X509KeyManager) getJavaDefaultKeyManager();
    }

    private static KeyManager getJavaDefaultKeyManager() throws Exception {
        KeyManagerFactory kmg = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
        kmg.init(null, null);
        // beriem prvy
        return kmg.getKeyManagers()[0];
    }

    @Override
    public String chooseClientAlias(String[] keyType, Principal[] issuers, Socket socket) {
        return javaDefaultKeyManager.chooseClientAlias(keyType, issuers, socket);
    }

    @Override
    public String chooseServerAlias(String keyType, Principal[] issuers, Socket socket) {
        return javaDefaultKeyManager.chooseServerAlias(keyType, issuers, socket);
    }

    @Override
    public X509Certificate[] getCertificateChain(String alias) {
        return javaDefaultKeyManager.getCertificateChain(alias);
    }

    @Override
    public String[] getClientAliases(String keyType, Principal[] issuers) {
        return javaDefaultKeyManager.getClientAliases(keyType, issuers);
    }

    @Override
    public PrivateKey getPrivateKey(String alias) {
        return javaDefaultKeyManager.getPrivateKey(alias);
    }

    @Override
    public String[] getServerAliases(String keyType, Principal[] issuers) {
        return javaDefaultKeyManager.getServerAliases(keyType, issuers);
    }
}

File: src/main/java/sk/hudak/prco/ssl/PrcoCustomKeyManager.kt
package sk.hudak.prco.ssl

import org.slf4j.LoggerFactory
import java.net.Socket
import java.security.Principal
import java.security.PrivateKey
import java.security.cert.X509Certificate
import javax.net.ssl.KeyManagerFactory
import javax.net.ssl.X509KeyManager

class PrcoCustomKeyManager @Throws(Exception::class)
constructor() : X509KeyManager {

    companion object {
        val log = LoggerFactory.getLogger(PrcoCustomKeyManager::class.java)
    }

    private val javaDefaultKeyManager: X509KeyManager

    init {
        log.debug("initializing")
        this.javaDefaultKeyManager = getJavaDefaultKeyManager()
    }

    @Throws(Exception::class)
    private fun getJavaDefaultKeyManager(): X509KeyManager {
        val kmg = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm())
        kmg.init(null, null)
        // beriem prvy
        return kmg.keyManagers[0] as X509KeyManager
    }

    override fun chooseClientAlias(keyType: Array<String>, issuers: Array<Principal>, socket: Socket): String {
        return javaDefaultKeyManager.chooseClientAlias(keyType, issuers, socket)
    }

    override fun chooseServerAlias(keyType: String, issuers: Array<Principal>, socket: Socket): String {
        return javaDefaultKeyManager.chooseServerAlias(keyType, issuers, socket)
    }

    override fun getCertificateChain(alias: String): Array<X509Certificate> {
        return javaDefaultKeyManager.getCertificateChain(alias)
    }

    override fun getClientAliases(keyType: String, issuers: Array<Principal>): Array<String> {
        return javaDefaultKeyManager.getClientAliases(keyType, issuers)
    }

    override fun getPrivateKey(alias: String): PrivateKey {
        return javaDefaultKeyManager.getPrivateKey(alias)
    }

    override fun getServerAliases(keyType: String, issuers: Array<Principal>): Array<String> {
        return javaDefaultKeyManager.getServerAliases(keyType, issuers)
    }
}
